File: ScapesCore/src/main/java/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteServer.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.chunk.terrain.infinite;

import java8.util.Optional;
import java8.util.function.Consumer;
import java8.util.function.Function;
import java8.util.stream.Collectors;
import java8.util.stream.Stream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.block.BlockType;
import org.tobi29.scapes.block.Update;
import org.tobi29.scapes.chunk.MobSpawner;
import org.tobi29.scapes.chunk.WorldServer;
import org.tobi29.scapes.chunk.generator.GeneratorOutput;
import org.tobi29.scapes.chunk.terrain.TerrainServer;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.io.tag.TagStructure;
import org.tobi29.scapes.engine.utils.math.FastMath;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;
import org.tobi29.scapes.engine.utils.math.vector.Vector2d;
import org.tobi29.scapes.engine.utils.math.vector.Vector2i;
import org.tobi29.scapes.engine.utils.math.vector.Vector3d;
import org.tobi29.scapes.engine.utils.profiler.Profiler;
import org.tobi29.scapes.engine.utils.task.Joiner;
import org.tobi29.scapes.engine.utils.task.TaskExecutor;
import org.tobi29.scapes.entity.server.EntityServer;
import org.tobi29.scapes.entity.server.MobPlayerServer;
import org.tobi29.scapes.server.format.TerrainInfiniteFormat;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ThreadLocalRandom;

public class TerrainInfiniteServer extends TerrainInfinite
        implements TerrainServer.TerrainMutable {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(TerrainInfiniteServer.class);
    protected final Object entityLock = new Object();
    private final WorldServer world;
    private final TerrainInfiniteFormat format;
    private final Queue<BlockChanges> blockChanges =
            new ConcurrentLinkedQueue<>();
    private final Queue<TerrainInfiniteChunkServer> chunkUnloadQueue =
            new ConcurrentLinkedQueue<>();
    private final TerrainInfiniteChunkManagerServer chunkManager;
    private final GeneratorOutput generatorOutput;
    private final Joiner updateJoiner, joiner;

    public TerrainInfiniteServer(WorldServer world, int zSize,
            TerrainInfiniteFormat format, TaskExecutor taskExecutor,
            BlockType voidBlock) {
        super(zSize, new TerrainInfiniteChunkManagerServer(), taskExecutor,
                voidBlock);
        this.world = world;
        this.format = format;
        chunkManager = (TerrainInfiniteChunkManagerServer) super.chunkManager;
        generatorOutput = new GeneratorOutput(zSize);
        Joiner loadJoiner = world.taskExecutor().runTask(joiner -> {
            List<Vector2i> requiredChunks = new ArrayList<>();
            List<Vector2i> loadingChunks = new ArrayList<>();
            while (!joiner.marked()) {
                Collection<MobPlayerServer> players = world.players();
                if (players.isEmpty()) {
                    Streams.of(chunkManager.iterator())
                            .forEach(chunkUnloadQueue::add);
                    removeChunks();
                    joiner.sleep(100);
                } else {
                    for (MobPlayerServer player : players) {
                        int xx = FastMath.floor(player.x() / 16.0);
                        int yy = FastMath.floor(player.y() / 16.0);
                        int loadingRadius =
                                (player.connection().loadingRadius() >> 4) + 3;
                        int loadingRadiusSqr = loadingRadius * loadingRadius;
                        for (int x = -loadingRadius; x <= loadingRadius; x++) {
                            int xxx = x + xx;
                            for (int y = -loadingRadius; y <= loadingRadius;
                                    y++) {
                                int yyy = y + yy;
                                if (xxx >= cxMin && xxx <= cxMax &&
                                        yyy >= cyMin &&
                                        yyy <= cyMax) {
                                    if (x * x + y * y < loadingRadiusSqr) {
                                        requiredChunks.add(new Vector2i(x + xx,
                                                y + yy));
                                    }
                                }
                            }
                        }
                    }
                    for (MobPlayerServer player : players) {
                        Vector2 loadArea = new Vector2d(player.x() / 16.0,
                                player.y() / 16.0);
                        Streams.of(requiredChunks)
                                .filter(pos -> !hasChunk(pos.intX(),
                                        pos.intY()))
                                .forEach(loadingChunks::add);
                        List<Vector2i> newChunks;
                        if (loadingChunks.size() > 64) {
                            newChunks = Streams.of(loadingChunks)
                                    .sorted((pos1, pos2) -> {
                                        double distance1 =
                                                FastMath.pointDistanceSqr(
                                                        loadArea, pos1);
                                        double distance2 =
                                                FastMath.pointDistanceSqr(
                                                        loadArea, pos2);
                                        return distance1 == distance2 ? 0 :
                                                distance1 > distance2 ? 1 : -1;
                                    }).limit(32).collect(Collectors.toList());
                        } else {
                            newChunks = Streams.of(loadingChunks).limit(32)
                                    .collect(Collectors.toList());
                        }
                        addChunks(newChunks);
                        Collection<TerrainInfiniteChunkServer> chunks =
                                chunkManager.iterator();
                        Streams.of(chunks)
                                .filter(TerrainInfiniteChunkServer::shouldPopulate)
                                .limit(32)
                                .forEach(TerrainInfiniteChunkServer::populate);
                        Streams.of(chunks)
                                .filter(TerrainInfiniteChunkServer::shouldFinish)
                                .forEach(TerrainInfiniteChunkServer::finish);
                        long time = System.currentTimeMillis() - 2000;
                        Streams.of(chunks)
                                .filter(chunk -> chunk.lastAccess() < time)
                                .filter(chunk -> !requiredChunks
                                        .contains(chunk.pos()))
                                .forEach(chunkUnloadQueue::add);
                        Streams.of(chunks).forEach(
                                TerrainInfiniteChunkServer::updateAdjacent);
                        if (removeChunks() && loadingChunks.isEmpty()) {
                            joiner.sleep(100);
                        }
                        loadingChunks.clear();
                    }
                    requiredChunks.clear();
                }
            }
        }, "Chunk-Loading");
        updateJoiner = world.taskExecutor().runTask(joiner -> {
            while (!joiner.marked()) {
                boolean idle = true;
                while (!blockChanges.isEmpty()) {
                    blockChanges.poll().run(this);
                    idle = false;
                }
                if (idle) {
                    joiner.sleep();
                }
            }
        }, "Chunk-Updating");
        joiner = new Joiner(loadJoiner, updateJoiner);
    }

    @Override
    public int sunLightReduction(int x, int y) {
        return (int) world.environment().sunLightReduction(x, y);
    }

    @Override
    public Optional<TerrainInfiniteChunkServer> chunkNoLoad(int x, int y) {
        return chunkManager.get(x, y);
    }

    @Override
    public Optional<TerrainInfiniteChunkServer> chunk(int x, int y) {
        Optional<TerrainInfiniteChunkServer> chunk = chunkManager.get(x, y);
        if (chunk.isPresent()) {
            chunk.get().accessed();
            return chunk;
        }
        return addChunk(x, y);
    }

    public boolean chunkS(int x, int y,
            Consumer<TerrainInfiniteChunkServer> consumer) {
        Optional<TerrainInfiniteChunkServer> chunk = chunk(x, y);
        if (chunk.isPresent()) {
            consumer.accept(chunk.get());
            return true;
        }
        return false;
    }

    public <R> Optional<R> chunkReturnS(int x, int y,
            Function<TerrainInfiniteChunkServer, R> consumer) {
        Optional<TerrainInfiniteChunkServer> chunk = chunk(x, y);
        if (chunk.isPresent()) {
            return Optional.of(consumer.apply(chunk.get()));
        }
        return Optional.empty();
    }

    public Optional<TerrainInfiniteChunkServer> addChunk(int x, int y) {
        return addChunks(Collections.singletonList(new Vector2i(x, y))).get(0);
    }

    public List<Optional<TerrainInfiniteChunkServer>> addChunks(
            List<Vector2i> positions) {
        List<Optional<TerrainInfiniteChunkServer>> chunks =
                new ArrayList<>(positions.size());
        List<Optional<TagStructure>> tagStructures =
                format.chunkTags(positions);
        for (int i = 0; i < positions.size(); i++) {
            Vector2i pos = positions.get(i);
            int x = pos.intX();
            int y = pos.intY();
            if (x < cxMin || x > cxMax || y < cyMin || y > cyMax) {
                chunks.add(Optional.empty());
                continue;
            }
            Optional<TerrainInfiniteChunkServer> chunk;
            synchronized (chunkManager) {
                chunk = chunkManager.get(x, y);
                if (!chunk.isPresent()) {
                    Optional<TagStructure> tagStructure = tagStructures.get(i);
                    TerrainInfiniteChunkServer chunk2;
                    if (tagStructure.isPresent()) {
                        try (Profiler.C ignored = Profiler.section("Load")) {
                            chunk2 = new TerrainInfiniteChunkServer(
                                    new Vector2i(x, y), this, zSize,
                                    tagStructure.get());
                        }
                    } else {
                        try (Profiler.C ignored = Profiler
                                .section("Generate")) {
                            chunk2 = new TerrainInfiniteChunkServer(
                                    new Vector2i(x, y), this, zSize,
                                    world.generator(), generatorOutput);
                        }
                    }
                    chunkManager.add(chunk2);
                    updateAdjacent(chunk2.x(), chunk2.y());
                    chunk = chunk2.optional();
                }
            }
            chunks.add(chunk);
        }
        return chunks;
    }

    @Override
    public WorldServer world() {
        return world;
    }

    @Override
    public void update(double delta, Collection<MobSpawner> spawners) {
        try (Profiler.C ignored = Profiler.section("Chunks")) {
            Streams.of(chunkManager.iterator())
                    .forEach(chunk -> chunk.updateServer(delta));
        }
        try (Profiler.C ignored = Profiler.section("Spawning")) {
            Random random = ThreadLocalRandom.current();
            for (MobSpawner spawner : spawners) {
                if (world.mobs(spawner.creatureType()) <
                        chunkManager.chunks() * spawner.mobsPerChunk()) {
                    for (TerrainInfiniteChunkServer chunk : chunkManager
                            .iterator()) {
                        if (random.nextInt(spawner.chunkChance()) == 0 &&
                                chunk.isLoaded()) {
                            for (int i = 0; i < spawner.spawnAttempts(); i++) {
                                int x = random.nextInt(16);
                                int y = random.nextInt(16);
                                int z = random.nextInt(chunk.zSize());
                                int xx = x + chunk.blockX();
                                int yy = y + chunk.blockY();
                                if (spawner.creatureType().doesDespawn()) {
                                    MobPlayerServer player =
                                            world.nearestPlayer(
                                                    new Vector3d(xx, yy, z));
                                    if (player == null) {
                                        continue;
                                    } else {
                                        double distance =
                                                FastMath.sqr(player.x() - xx) +
                                                        FastMath.sqr(
                                                                player.y() -
                                                                        yy) +
                                                        FastMath.sqr(
                                                                player.z() - z);
                                        if (distance > 9216.0 ||
                                                distance < 256.0) {
                                            continue;
                                        }
                                    }
                                }
                                if (spawner.canSpawn(this, xx, yy, z)) {
                                    EntityServer entity =
                                            spawner.spawn(this, xx, yy, z);
                                    entity.onSpawn();
                                    world.addEntity(entity);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @Override
    public void queue(BlockChanges blockChanges) {
        this.blockChanges.add(blockChanges);
        updateJoiner.wake();
    }

    @Override
    public void addDelayedUpdate(Update update) {
        chunkS(update.x() >> 4, update.y() >> 4,
                chunk -> chunk.addDelayedUpdate(update));
    }

    @Override
    public boolean hasDelayedUpdate(int x, int y, int z,
            Class<? extends Update> clazz) {
        if (z < 0 || z >= zSize) {
            return false;
        }
        return chunkReturnS(x >> 4, y >> 4,
                chunk -> chunk.hasDelayedUpdate(x, y, z, clazz)).orElse(false);
    }

    @Override
    public boolean isBlockSendable(MobPlayerServer player, int x, int y, int z,
            boolean chunkContent) {
        if (z < 0 || z >= zSize) {
            return false;
        }
        int loadingRadius = player.connection().loadingRadius() >> 4;
        x = x >> 4;
        y = y >> 4;
        int x2 = FastMath.floor(player.x() / 16.0);
        int y2 = FastMath.floor(player.y() / 16.0);
        Optional<TerrainInfiniteChunkServer> chunk = chunkNoLoad(x, y);
        int dis = FastMath.abs(x - x2);
        dis = FastMath.max(dis, FastMath.abs(y - y2));
        if (chunk.isPresent()) {
            if (dis > loadingRadius ||
                    !chunk.get().isSendable() && chunkContent) {
                return false;
            }
        } else {
            if (dis > loadingRadius) {
                return false;
            }
        }
        return true;
    }

    @Override
    public boolean addEntity(EntityServer entity) {
        synchronized (entityLock) {
            return addEntity(entity, freeEntityID());
        }
    }

    @Override
    public void addPlayer(MobPlayerServer player) {
        synchronized (entityLock) {
            player.setEntityID(freeEntityID());
            world.addPlayer(player);
        }
    }

    @Override
    public boolean removeEntity(EntityServer entity) {
        if (entity != null) {
            int x = FastMath.floor(entity.x()) >> 4;
            int y = FastMath.floor(entity.y()) >> 4;
            if (chunkReturnS(x, y, chunk -> chunk.removeEntity(entity))
                    .orElse(false)) {
                return true;
            }
            for (TerrainInfiniteChunkServer chunk : chunkManager.iterator()) {
                if (chunk.removeEntity(entity)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public Optional<EntityServer> entity(int id) {
        Optional<MobPlayerServer> player = Streams.of(world.players())
                .filter(entity -> entity.entityID() == id).findAny();
        if (player.isPresent()) {
            return Optional.of(player.get());
        }
        return Streams.of(chunkManager.iterator())
                .map(chunk -> chunk.entity(id)).filter(Optional::isPresent)
                .map(Optional::get).findAny();
    }

    @Override
    public void entities(Consumer<Stream<EntityServer>> consumer) {
        for (TerrainInfiniteChunkServer chunk : chunkManager.iterator()) {
            consumer.accept(chunk.entities());
        }
    }

    @Override
    public void entities(int x, int y, int z,
            Consumer<Stream<EntityServer>> consumer) {
        chunkS(x >> 4, y >> 4, chunk -> consumer.accept(chunk.entities()
                .filter(entity -> FastMath.floor(entity.x()) == x)
                .filter(entity -> FastMath.floor(entity.y()) == y)
                .filter(entity -> FastMath.floor(entity.z()) == z)));
    }

    @Override
    public void entitiesAtLeast(int minX, int minY, int minZ, int maxX,
            int maxY, int maxZ, Consumer<Stream<EntityServer>> consumer) {
        int minCX = minX >> 4;
        int minCY = minY >> 4;
        int maxCX = maxX >> 4;
        int maxCY = maxY >> 4;
        for (int yy = minCY; yy <= maxCY; yy++) {
            for (int xx = minCX; xx <= maxCX; xx++) {
                chunkNoLoad(xx, yy)
                        .ifPresent(chunk -> consumer.accept(chunk.entities()));
            }
        }
    }

    @Override
    public void dispose() {
        joiner.join();
        lighting.dispose();
        Streams.of(chunkManager.iterator()).forEach(chunkUnloadQueue::add);
        removeChunks();
        format.dispose();
    }

    public boolean addEntity(EntityServer entity, int id) {
        entity.setEntityID(id);
        int x = FastMath.floor(entity.x()) >> 4;
        int y = FastMath.floor(entity.y()) >> 4;
        return chunkS(x, y, chunk -> chunk.addEntity(entity));
    }

    protected int freeEntityID() {
        Random random = ThreadLocalRandom.current();
        while (true) {
            int i = random.nextInt(Integer.MAX_VALUE);
            if (!world.entity(i).isPresent()) {
                return i;
            }
        }
    }

    @Override
    public void type(int x, int y, int z, BlockType type) {
        if (z < 0 || z >= zSize) {
            return;
        }
        chunk(x >> 4, y >> 4, chunk -> chunk.blockTypeG(x, y, z, type));
    }

    @Override
    public void data(int x, int y, int z, int data) {
        if (z < 0 || z >= zSize) {
            return;
        }
        chunk(x >> 4, y >> 4, chunk -> chunk.dataG(x, y, z, data));
    }

    @Override
    public void typeData(int x, int y, int z, BlockType block, int data) {
        if (z < 0 || z >= zSize) {
            return;
        }
        chunk(x >> 4, y >> 4, chunk -> chunk.typeDataG(x, y, z, block, data));
    }

    public void updateAdjacent(int x, int y) {
        for (int xx = -1; xx <= 1; xx++) {
            int xxx = xx + x;
            for (int yy = -1; yy <= 1; yy++) {
                if (xx != 0 || yy != 0) {
                    chunkManager.get(xxx, yy + y).ifPresent(
                            TerrainInfiniteChunkServer::updateAdjacent);
                }
            }
        }
    }

    public boolean checkBorder(TerrainInfiniteChunkServer chunk, int radius) {
        for (int x = -radius; x <= radius; x++) {
            int xx = chunk.x() + x;
            for (int y = -radius; y <= radius; y++) {
                int yy = chunk.y() + y;
                if (xx >= cxMin && xx <= cxMax && yy >= cyMin && yy <= cyMax) {
                    if (!chunkNoLoad(xx, yy).isPresent()) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public boolean checkLoaded(TerrainInfiniteChunkServer chunk, int radius) {
        for (int x = -radius; x <= radius; x++) {
            int xx = chunk.x() + x;
            for (int y = -radius; y <= radius; y++) {
                int yy = chunk.y() + y;
                if (xx >= cxMin && xx <= cxMax && yy >= cyMin && yy <= cyMax) {
                    Optional<TerrainInfiniteChunkServer> check =
                            chunkNoLoad(xx, yy);
                    if (!check.isPresent() || !check.get().isLoaded()) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    private boolean removeChunks() {
        List<Pair<Vector2i, TagStructure>> chunks = new ArrayList<>();
        while (!chunkUnloadQueue.isEmpty()) {
            synchronized (this) {
                removeChunk(chunkUnloadQueue.poll()).ifPresent(chunks::add);
            }
        }
        if (chunks.isEmpty()) {
            return true;
        }
        try {
            format.putChunkTags(chunks);
        } catch (IOException e) {
            LOGGER.error("Failed to store chunks:", e);
        }
        return false;
    }

    private Optional<Pair<Vector2i, TagStructure>> removeChunk(
            TerrainInfiniteChunkServer chunk) {
        int x = chunk.x();
        int y = chunk.y();
        chunkManager.remove(x, y);
        TagStructure tagStructure = chunk.dispose();
        while (chunkUnloadQueue.remove(chunk)) {
            LOGGER.warn("Chunk queued for unloading twice!");
        }
        updateAdjacent(chunk.x(), chunk.y());
        return Optional.of(new Pair<>(new Vector2i(x, y), tagStructure));
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteServer.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.chunk.terrain.infinite

import kotlinx.coroutines.experimental.CoroutineName
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.launch
import org.tobi29.coroutines.Timer
import org.tobi29.coroutines.loop
import org.tobi29.logging.KLogging
import org.tobi29.scapes.block.BlockType
import org.tobi29.scapes.block.Update
import org.tobi29.scapes.chunk.MobSpawner
import org.tobi29.scapes.chunk.WorldServer
import org.tobi29.scapes.chunk.generator.ChunkGenerator
import org.tobi29.scapes.chunk.generator.ChunkPopulator
import org.tobi29.scapes.chunk.terrain.TerrainMutableServer
import org.tobi29.scapes.chunk.terrain.TerrainServer
import org.tobi29.math.threadLocalRandom
import org.tobi29.math.vector.Vector2i
import org.tobi29.math.vector.Vector3d
import org.tobi29.math.vector.distanceSqr
import org.tobi29.io.IOException
import org.tobi29.profiler.profilerSection
import org.tobi29.scapes.entity.server.EntityServer
import org.tobi29.scapes.entity.server.MobPlayerServer
import org.tobi29.scapes.server.format.TerrainInfiniteFormat
import org.tobi29.scapes.terrain.TerrainChunk
import org.tobi29.scapes.terrain.infinite.TerrainInfiniteChunkManagerDynamic
import org.tobi29.io.tag.TagMap
import org.tobi29.stdex.ThreadLocal
import org.tobi29.stdex.atomic.AtomicBoolean
import org.tobi29.stdex.math.floorToInt
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.TimeUnit
import kotlin.math.abs
import kotlin.math.max

class TerrainInfiniteServer(override val world: WorldServer,
                            zSize: Int,
                            private val format: TerrainInfiniteFormat,
                            override val generator: ChunkGenerator,
                            internal val populators: Array<ChunkPopulator>,
                            air: BlockType) : TerrainInfinite<EntityServer, TerrainInfiniteChunkServer>(
        zSize, world.taskExecutor, air, world.registry,
        TerrainInfiniteChunkManagerDynamic<TerrainInfiniteChunkServer>()),
        TerrainServer {
    // TODO: Port away
    private val chunkUnloadQueue = ConcurrentLinkedQueue<TerrainInfiniteChunkServer>()
    private var loadJob: Pair<Job, AtomicBoolean>? = null

    init {
        val stop = AtomicBoolean(false)
        loadJob = launch(world.taskExecutor + CoroutineName("Chunk-Loading")) {
            val loader = TerrainInfiniteChunkLoader(loadedChunks(), cxMin,
                    cxMax, cyMin, cyMax)

            Timer().apply { init() }.loop(Timer.toDiff(10.0),
                    { delay(it, TimeUnit.NANOSECONDS) }) {
                if (stop.get()) return@loop false

                while (true) {
                    for (player in world.players()) {
                        val pos = player.getCurrentPos()
                        val xx = pos.x.floorToInt() shr 4
                        val yy = pos.y.floorToInt() shr 4
                        val loadingRadius = (player.connection().loadingRadius() shr 4) + 2
                        loader.requireCircle(xx, yy, loadingRadius)
                    }

                    val activeRemove = removeChunks(loader.outsideChunks)
                    val activeAdd = addChunks(loader.requiredChunks {
                        !chunkManager.has(it.x, it.y)
                    }.map { it.now() }.toList()).isNotEmpty()
                    chunkManager.stream().forEach { it.updateAdjacent() }
                    chunkManager.stream().filter { it.shouldPopulate() }
                            .forEach { it.populate() }
                    chunkManager.stream().filter { it.shouldFinish() }
                            .forEach(TerrainInfiniteChunkServer::finish)
                    val active = activeAdd || activeRemove
                    loader.reset()

                    if (!active) break
                }

                true
            }
        } to stop
    }

    override fun getThreadContext(): TerrainInfiniteServerSection = SECTION.get()

    override fun sunLightReduction(x: Int,
                                   y: Int): Int {
        return world.environment.sunLightReduction(x.toDouble(),
                y.toDouble()).toInt()
    }

    override fun addEntity(entity: EntityServer,
                           spawn: Boolean): Boolean {
        val pos = entity.getCurrentPos()
        val x = pos.x.floorToInt() shr 4
        val y = pos.y.floorToInt() shr 4
        return chunk(x, y, { chunk ->
            chunk.addEntity(entity)
            true
        }) ?: false
    }

    override fun entityAdded(entity: EntityServer,
                             spawn: Boolean) {
        super.entityAdded(entity, spawn)
        world.entityAdded(entity, spawn)
    }

    override fun entityRemoved(entity: EntityServer) {
        super.entityRemoved(entity)
        world.entityRemoved(entity)
    }

    override fun addChunk(x: Int,
                          y: Int): TerrainInfiniteChunkServer? {
        return addChunks(listOf(Vector2i(x, y)))[0]
    }

    fun addChunks(
            positions: List<Vector2i>): List<TerrainInfiniteChunkServer?> {
        val chunks = ArrayList<TerrainInfiniteChunkServer?>(
                positions.size)
        val chunkMaps = format.chunkTags(positions)
        for (i in positions.indices) {
            val pos = positions[i]
            val x = pos.x
            val y = pos.y
            if (x < cxMin || x > cxMax || y < cyMin || y > cyMax) {
                chunks.add(null)
                continue
            }
            chunks.add(synchronized(chunkManager) {
                var chunk = chunkManager[x, y]
                if (chunk == null) {
                    val map = chunkMaps[i]
                    chunk = TerrainInfiniteChunkServer(Vector2i(x, y), this,
                            zSize)
                    if (map == null) {
                        chunk.generate(generator)
                    } else {
                        chunk.load(map)
                    }
                    chunkManager.add(chunk)
                    updateAdjacent(chunk.pos.x, chunk.pos.y)
                }
                chunk
            })
        }
        return chunks
    }

    override fun update(delta: Double,
                        spawners: Collection<MobSpawner>) {
        profilerSection("Chunks") {
            chunkManager.stream().forEach { it.updateServer(delta) }
        }
        profilerSection("Spawning") {
            val random = threadLocalRandom()
            for (spawner in spawners) {
                if (world.mobs(
                        spawner.creatureType()) < chunkManager.chunks * spawner.mobsPerChunk()) {
                    for (chunk in chunkManager.stream()) {
                        if (random.nextInt(
                                spawner.chunkChance()) == 0 && chunk.isLoaded) {
                            for (i in 0 until spawner.spawnAttempts()) {
                                val x = random.nextInt(chunk.size.x)
                                val y = random.nextInt(chunk.size.y)
                                val z = random.nextInt(chunk.size.z)
                                val xx = x + chunk.posBlock.x
                                val yy = y + chunk.posBlock.y
                                if (spawner.creatureType().doesDespawn()) {
                                    val player = world.nearestPlayer(
                                            Vector3d(xx.toDouble(),
                                                    yy.toDouble(),
                                                    z.toDouble()))
                                    if (player == null) {
                                        continue
                                    } else {
                                        val distance = player.getCurrentPos() distanceSqr Vector3d(
                                                xx + 0.5, yy + 0.5, z + 0.5)
                                        if (distance > 9216.0 || distance < 256.0) {
                                            continue
                                        }
                                    }
                                }
                                if (spawner.canSpawn(this, xx, yy, z)) {
                                    val entity = spawner.spawn(this, xx, yy, z)
                                    world.addEntityNew(entity)
                                }
                            }
                        }
                    }
                }
            }
        }
        profilerSection("ChunkManager") {
            chunkManager.update()
        }
    }

    override fun <R> modify(x: Int,
                            y: Int,
                            z: Int,
                            dx: Int,
                            dy: Int,
                            dz: Int,
                            block: (TerrainMutableServer) -> R): R {
        val section = SECTION.get()
        profilerSection("Terrain-Modify-Init") {
            section.init(this, x, y, z, dx, dy, dz)
            section.lockChunks()
        }
        try {
            profilerSection("Terrain-Modify-Block") {
                return block(section)
            }
        } finally {
            profilerSection("Terrain-Modify-Flush") {
                section.unlockChunks()
                section.flushPackets()
                section.flushUpdates()
                section.clear()
            }
        }
    }

    override fun hasDelayedUpdate(x: Int,
                                  y: Int,
                                  z: Int,
                                  clazz: Class<out Update>): Boolean {
        if (z < 0 || z >= zSize) {
            return false
        }
        return chunkS(x shr 4, y shr 4) {
            it.hasDelayedUpdate(x, y, z, clazz)
        } ?: false
    }

    override fun isBlockSendable(player: MobPlayerServer,
                                 x: Int,
                                 y: Int,
                                 z: Int,
                                 chunkContent: Boolean): Boolean {
        if (z < 0 || z >= zSize) {
            return false
        }
        val loadingRadius = player.connection().loadingRadius() shr 4
        val xx = x shr 4
        val yy = y shr 4
        val pos = player.getCurrentPos()
        val x2 = pos.x.floorToInt() shr 4
        val y2 = pos.y.floorToInt() shr 4
        val chunk = chunkNoLoad(xx, yy)
        val dis = max(abs(xx - x2), abs(yy - y2))
        if (chunk != null) {
            if (dis > loadingRadius || !chunk.isSendable && chunkContent) {
                return false
            }
        } else {
            if (dis > loadingRadius) {
                return false
            }
        }
        return true
    }

    override fun chunks(consumer: (TerrainChunk) -> Unit) {
        chunkManager.stream().filter { it.isLoaded }.forEach { consumer(it) }
    }

    override suspend fun dispose() {
        loadJob?.let { (job, stop) ->
            stop.set(true)
            job.join()
        }
        lighting.dispose()
        chunkManager.stream().forEach { chunkUnloadQueue.add(it) }
        removeChunks()
        format.dispose()
    }

    fun updateAdjacent(x: Int,
                       y: Int) {
        for (xx in -1..1) {
            val xxx = xx + x
            for (yy in -1..1) {
                if (xx != 0 || yy != 0) {
                    chunkManager[xxx, yy + y]?.updateAdjacent()
                }
            }
        }
    }

    fun checkBorder(chunk: TerrainInfiniteChunkServer,
                    radius: Int): Boolean {
        for (x in -radius..radius) {
            val xx = chunk.pos.x + x
            for (y in -radius..radius) {
                val yy = chunk.pos.y + y
                if (xx in cxMin..cxMax && yy >= cyMin && yy <= cyMax) {
                    if (chunkNoLoad(xx, yy) == null) {
                        return false
                    }
                }
            }
        }
        return true
    }

    fun checkLoaded(chunk: TerrainInfiniteChunkServer,
                    radius: Int): Boolean {
        for (x in -radius..radius) {
            val xx = chunk.pos.x + x
            for (y in -radius..radius) {
                val yy = chunk.pos.y + y
                if (xx in cxMin..cxMax && yy >= cyMin && yy <= cyMax) {
                    val check = chunkNoLoad(xx, yy)
                    if (check == null || !check.isLoaded) {
                        return false
                    }
                }
            }
        }
        return true
    }

    private fun removeChunks(): Boolean {
        val chunks = ArrayList<Pair<Vector2i, TagMap>>()
        while (!chunkUnloadQueue.isEmpty()) {
            synchronized(this) {
                chunkUnloadQueue.poll()?.let { chunks.add(removeChunk(it)) }
            }
        }
        if (chunks.isEmpty()) {
            return true
        }
        try {
            format.putChunkTags(chunks)
        } catch (e: IOException) {
            logger.error(e) { "Failed to store chunks" }
        }
        return false
    }

    private fun removeChunks(oi: Sequence<TerrainInfiniteChunkServer>): Boolean {
        val chunks = ArrayList<Pair<Vector2i, TagMap>>()
        synchronized(this) {
            oi.forEach { chunks.add(removeChunk(it)) }
        }
        if (chunks.isEmpty()) {
            return false
        }
        try {
            format.putChunkTags(chunks)
        } catch (e: IOException) {
            logger.error(e) { "Failed to store chunks" }
        }
        return true
    }

    private fun removeChunk(
            chunk: TerrainInfiniteChunkServer): Pair<Vector2i, TagMap> {
        val x = chunk.pos.x
        val y = chunk.pos.y
        chunkManager.remove(x, y)
        val tagStructure = chunk.disposeAndWrite()
        while (chunkUnloadQueue.remove(chunk)) {
            logger.warn { "Chunk queued for unloading twice!" }
        }
        updateAdjacent(chunk.pos.x, chunk.pos.y)
        return Pair(Vector2i(x, y), tagStructure)
    }

    companion object : KLogging() {
        val SECTION = ThreadLocal { TerrainInfiniteServerSection() }
    }
}
